pootle/apps/virtualfolder/receivers.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) Pootle contributors.
+#
+# This file is a part of the Pootle project. It is distributed under the GPL3
+# or later license. See the LICENSE file for a copy of the license and the
+# AUTHORS file for copyright and authorship information.
+
+from django.db.models.signals import post_save, pre_save
+from django.dispatch import receiver
+
+from pootle_store.models import Unit
+from pootle_store.util import OBSOLETE
+
+from .models import VirtualFolder, VirtualFolderTreeItem
+from .signals import vfolder_post_save
+
+
+def update_vfolder_tree(vf, store):
+    # Create missing VirtualFolderTreeItem tree structure after
+    # adding this new unit.
+    vfolder_treeitem, created = (
+        VirtualFolderTreeItem.objects.get_or_create(
+            directory=store.parent, vfolder=vf))
+
+    if not created:
+        # The VirtualFolderTreeItem already existed, so
+        # calculate again the stats up to the root.
+        vfolder_treeitem.update_all_cache()
+
+
+def add_unit_to_vfolders(unit):
+    pootle_path = unit.store.pootle_path
+
+    for vf in VirtualFolder.objects.iterator():
+        unit_added = False
+        for location in vf.all_locations:
+            if not pootle_path.startswith(location):
+                continue
+
+            for filename in vf.filter_rules.split(","):
+                if pootle_path == "".join([location, filename]):
+                    vf.units.add(unit)
+                    unit_added = True
+                    break
+
+            if unit_added:
+                break
+
+        if unit_added:
+            update_vfolder_tree(vf, unit.store)
+
+
+@receiver(pre_save, sender=VirtualFolder)
+def vfolder_unit_priority_presave_handler(sender, instance, **kwargs):
+
+    if instance.id is None:
+        return
+
+    original = VirtualFolder.objects.get(pk=instance.pk)
```

Comment:
Since this is `pre_save` wouldn't `instance` lack the changes we want to avoid by doing this?